src/communication/distributed_context.hpp

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include <string>
+
+#include <spike.hpp>
+#include <communication/gathered_vector.hpp>
+#include <util/pp_util.hpp>
+
+#if defined(ARB_HAVE_MPI)
+#   include "mpi_context.hpp"
+#endif
+#include "serial_context.hpp"
+
+
+namespace arb {
+
+#define PUBLIC_COLLECTIVES(T) \
+    T min(T value) const { return impl_->min(value); }\
+    T max(T value) const { return impl_->max(value); }\
+    T sum(T value) const { return impl_->sum(value); }\
+    std::vector<T> gather(T value, int root) const { return impl_->gather(value, root); }
+
+#define INTERFACE_COLLECTIVES(T) \
+    virtual T min(T value) const = 0;\
+    virtual T max(T value) const = 0;\
+    virtual T sum(T value) const = 0;\
+    virtual std::vector<T> gather(T value, int root) const = 0;
+
+#define WRAP_COLLECTIVES(T) \
+    T min(T value) const override { return wrapped.min(value); }\
+    T max(T value) const override { return wrapped.max(value); }\
+    T sum(T value) const override { return wrapped.sum(value); }\
+    std::vector<T> gather(T value, int root) const override { return wrapped.gather(value, root); }
+
+#define COLLECTIVE_TYPES float, double, int, std::uint32_t, std::uint64_t
+
+class distributed_context {
+public:
+    using spike_vector = std::vector<arb::spike>;
+
+    // default constructor uses a serial context
+    distributed_context(): distributed_context(serial_context()) {}
+
+    template <typename Impl>
+    distributed_context(Impl&& impl):
+        impl_(new wrap<Impl>(std::forward<Impl>(impl)))
+    {}
+
+    distributed_context(distributed_context&& other) = default;
+    distributed_context& operator=(distributed_context&& other) = default;
+
+    gathered_vector<arb::spike> gather_spikes(const spike_vector& local_spikes) const {
+        return impl_->gather_spikes(local_spikes);
+    }
+
+    int id() const {
+        return impl_->id();
+    }
+
+    int size() const {
+        return impl_->size();
+    }
+
+    void barrier() const {
+        impl_->barrier();
+    }
+
+    std::string name() const {
+        return impl_->name();
+    }
+
+    PP_FOREACH(PUBLIC_COLLECTIVES, COLLECTIVE_TYPES);
```

Comment:
Similarly, we should probably rename this macro to `ARB_PP_FOREACH` in order to reduce the chance of macro name collisions.

Reply:
Done.